### PR TITLE
Added the ability to terminate a run

### PR DESCRIPTION
**Pull request by Ark-kun**
---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/528)

### CODE REVIEW

File: backend/src/common/client/api_server/run_client_fake.go [hunk 1]
```diff
@@ -74,3 +74,14 @@ func (c *RunClientFake) ListAll(params *runparams.ListRunsParams, maxResultSize
 	[]*runmodel.APIRun, error) {
 	return listAllForRun(c, params, maxResultSize)
 }
+
+func (c *RunClientFake) Terminate(params *runparams.TerminateRunParams) error {
```
Comment (yebrahim) [inline backend/src/common/client/api_server/run_client_fake.go]: We'll need an integration test for this, can you add a TODO and open an issue to track it?

File: backend/src/apiserver/storage/run_store.go [hunk 2]
```diff
@@ -401,3 +409,51 @@ func (s *RunStore) toRunMetadatas(models []model.ListableDataModel) []model.Run
...
+	_, err = s.db.Exec(`
+		UPDATE run_details
+		SET Conditions = ?,
+		WHERE UUID = ? AND Conditions = ?`,
+		"Terminating:", runId, workflowapi.NodeRunning+":")
+	if err != nil {
+		return util.NewInternalServerError(err, "Failed to start terminating the run: %s", err.Error())
+	}
+
+	workflowInterface, err := client.CreateWorkflowClient("default")
```
[the remark this answers is not in the document]
Comment (Ark-kun) [reply backend/src/apiserver/storage/run_store.go]: Thanks for finding these problems.
Please note that these commits were made back in September. They might be correct at that time, but not compatible with later changes.

File: backend/src/apiserver/server/run_server.go [hunk 3]
```diff
@@ -122,6 +122,14 @@ func (s *RunServer) validateCreateRunRequest(request *api.CreateRunRequest) erro
 	return nil
 }
 
+func (s *RunServer) TerminateRun(ctx context.Context, request *api.TerminateRunRequest) (*empty.Empty, error) {
```
[the remark this answers is not in the document]
Comment (Ark-kun) [reply backend/src/apiserver/server/run_server.go]: I copied the style from `func (s *JobServer) DisableJob(ctx context.Context, request *api.DisableJobRequest) (*empty.Empty, error) {`

AFAIK, there was some compilation problem if the result was just an error. I'll try it again in the new code base.

**Comment by vicaire** [comment]
(I am assuming this is WIP because of the title. Let me know when ready).

File: backend/api/go_http_client/run_client/run_service/run_service_client.go [hunk 4]
```diff
@@ -212,6 +212,35 @@ func (a *Client) ReportRunMetrics(params *ReportRunMetricsParams, authInfo runti
 
 }
 
+/*
+TerminateRun terminate run API
+*/
+func (a *Client) TerminateRun(params *TerminateRunParams, authInfo runtime.ClientAuthInfoWriter) (*TerminateRunOK, error) {
```
Comment (vicaire) [inline backend/api/go_http_client/run_client/run_service/run_service_client.go]: For each API that you call, please have an abstract client with a fake implementation (for tests) and a real implementation (for production). 

See this example for the persistence agent: 
https://github.com/kubeflow/pipelines/tree/master/backend/src/agent/persistence/client

File: backend/src/apiserver/storage/run_store.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ark-kun) [reply backend/src/apiserver/storage/run_store.go]: Done.

File: backend/src/apiserver/server/run_server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ark-kun) [reply backend/src/apiserver/server/run_server.go]: The code generated from .proto has this signature:
https://github.com/kubeflow/pipelines/blob/a374585cac97e4fe60bb53d93c19f365516e5f44/backend/api/go_client/run.pb.go#L1197

All of the methods in run.proto have similar interface:
https://github.com/kubeflow/pipelines/blob/a374585cac97e4fe60bb53d93c19f365516e5f44/backend/api/run.proto#L81

Do you think this should be changed? Maybe we should change this everywhere.

[the remark this answers is not in the document]
Comment (yebrahim) [reply backend/src/apiserver/server/run_server.go]: Yes but I'll let @IronPan weigh in on the style, there might be a good reason behind it.

File: backend/src/apiserver/server/run_server.go [hunk 5]
```diff
@@ -134,6 +134,14 @@ func (s *RunServer) validateCreateRunRequest(request *api.CreateRunRequest) erro
...
+	err := s.resourceManager.TerminateRun(request.RunId)
+	if err != nil {
```
Comment (yebrahim) [inline backend/src/apiserver/server/run_server.go]: Replace this method's body with just a call to TerminateRun?
```suggestion
	return nil, s.resourceManager.TerminateRun(request.RunId)
```

Comment (rileyjbauer) [reply backend/src/apiserver/server/run_server.go]: I think the success return value needs to be `return &empty.Empty{}, nil`, not `nil, nil`

File: backend/api/go_http_client/run_client/run_service/run_service_client.go (hunk 4 above)
Comment (rileyjbauer) [reply backend/api/go_http_client/run_client/run_service/run_service_client.go]: `workflow_fake.go` has been updated to fake this behavior.
see: https://github.com/kubeflow/pipelines/pull/528/files#diff-b43f69f9e25073a72d6ebec688536c24

File: backend/src/cmd/ml/cmd/run.go [hunk 6]
```diff
@@ -95,6 +95,40 @@ func NewRunListCmd(root *RootCommand, pageSize int32) *cobra.Command {
...
+		err   error
+	)
+	var command = &cobra.Command{
+		Use:   "terminate ID",
+		Short: "Terminate a run",
+
+		// Validation
+		Args: func(cmd *cobra.Command, args []string) error {
+			runID, err = ValidateSingleString(args, "ID")
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (rileyjbauer) [reply backend/src/cmd/ml/cmd/run.go]: Done

File: backend/src/apiserver/server/run_server.go (hunk 5 above)
Comment (Ark-kun) [reply backend/src/apiserver/server/run_server.go]: That's the reason, yes.

File: backend/api/go_http_client/run_client/run_service/run_service_client.go (hunk 4 above)
Comment (vicaire) [reply backend/api/go_http_client/run_client/run_service/run_service_client.go]: Thank you Riley. Looks like this PR is still WIP according to the title. Let me know when it is ready for review. Thanks!

**Pull request merged**